### ISSUE TITLE
auto-resolve all/most merge conflicts

### ISSUE THREAD

**Issue by seansfkelley** [sdispater/poetry#496]
- [x] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.

## Issue

When two devs install dependencies on separate branches, it is very easy to end up merge-conflicted, in particular, the `metadata.content-hash` key often changes. It is very unclear how to resolve this manually, so I often delete the lockfile (or perhaps just that key) and rebuild it and, basically, hope that it comes out the same.

It seems like in some scenarios that merge conflicts could be resolved automatically based on pyproject. https://github.com/yarnpkg/yarn/ does this, for instance.

**Comment by simonpercivall** [comment]
+1 Resolving the `content-hash` and trivial conflicts in the individual hashes section would be very welcome.

**Comment by jimbo1qaz** [comment]
Workaround:
- remove content-hash line
- `poetry add pathlib2`
- `poetry remove pathlib2`.

Will this issue be resolved soon?

Does a missing content-hash hurt security or not?

**Comment by seansfkelley** [comment]
Yarn's implementation of this is way less insane than I was expecting:

1. Extract the two versions of the conflicted file.
    https://github.com/yarnpkg/yarn/blob/a7334da31bf783af7a3efab451589fe7ac34c748/src/lockfile/parse.js#L397

2. Blindly try to parse the files, and if that's successful, shallowly merge them.
    https://github.com/yarnpkg/yarn/blob/a7334da31bf783af7a3efab451589fe7ac34c748/src/lockfile/parse.js#L399

If the merge conflict resulted in a syntax error, it fails. Yarn's lockfile structure is designed to make this easier: it's flat, it's sorted alphabetically and node is okay with duplicated dependencies in the tree.

**Comment by itajaja** [comment]
is `content-hash` mandatory though? the main problem I am encountering is that dependabot will have to ALWAYS rebase all the PRs, which could be avoided if `content-hash` was not in poetry.lock. in theory shouldn't be necessary, as it can always be computed by the other hashes, right>

**Comment by chdsbd** [comment]
I think this is still a valid issue and should not be closed.

**Comment by cjolowicz** [comment]
I have written a stand-alone script as a stopgap. It only goes half the way, resolving the merge conflict for `metadata.content-hash`. The remaining conflicts, if any, should be quite trivial to merge manually. Use at your own risk.

https://github.com/cjolowicz/scripts/blob/master/python/poetry-merge-lock.py

**Comment by cjolowicz** [comment]
Here is a stand-alone tool which should handle most merge conflicts in the lock file:

https://github.com/cjolowicz/poetry-merge-lock

This tool is in early development. If you're interested in trying it out, please let me know on its issue tracker if you encounter any problems.

https://github.com/cjolowicz/poetry-merge-lock/issues

@sdispater, @finswimmer : Would you be interested in a PR to add this as a Poetry command?

**Comment by finswimmer** [comment]
Hello @cjolowicz ,

your contribution is very welcome. I find this feature very useful. :+1: 

However, I cannot promise if and when this can be included. Including new features is decided by @sdispater .

But please go on!

fin swimmer

**Comment by donbowman** [comment]
as long as there is a line `content-hash` we will never be able to conveniently use dependabot. When dependabot runs, it creates 1 branch and merge request per update, w/ the idea that you let your CI run, and then auto-merge.
However, because of `content-hash`, with poetry, every single one of these branches has a merge conflict and must be manually dealt with, increasing the human time from a few seconds to ~10-minutes per, and, causing another run of the CI to be required.

The repository I just tried this on, 5 MR were created. If it uses tox, it takes no human time and 5 CI runs.
If I use poetry it takes 10 CI runs, and a manual clone/checkout/rebase/poetry update hash somehow/push, which took me more than an hour.

Multiple this by my 30 repos, and make it a daily thing to have to deal with and it becomes intractable.

The solution cannot be in a new poetry command to resolve the lock, that would still require all the manual work.
Instead, we need to be able to merge poetry.lock files that are not otherwise conflicting.

Beacuse all these MR are created at once, in parallel, we cannot teach dependabot to do it either, at the time of each MR they all come from the same spot on master so there is no conflict, its only when the first is accepted that we need to rebase the 2nd, and when it is accepted we need to rebase (twice) the 3rd, and so on.

Can we not make the issuance of `content-hash` optional?

event without dependabot its troublesome.

**Comment by cjolowicz** [comment]
@donbowman Doesn't Dependabot rebase your PRs automatically if they conflict? This is what happens for me. I only need to resolve conflicts when rebasing or cherry-picking my own commits, in which case [poetry-merge-lock](https://github.com/cjolowicz/poetry-merge-lock) mostly works fine. I sometimes follow it by `poetry add insecure-package && poetry remove insecure-package` to ensure that the lock file is up-to-date.

**Comment by donbowman** [comment]
> @donbowman Doesn't Dependabot rebase your PRs automatically if they conflict? This is what happens for me. I only need to resolve conflicts when rebasing or cherry-picking my own commits, in which case [poetry-merge-lock](https://github.com/cjolowicz/poetry-merge-lock) mostly works fine. I sometimes follow it by `poetry add insecure-package && poetry remove insecure-package` to ensure that the lock file is up-to-date. ([insecure-package](https://pypi.org/project/insecure-package/) is just an empty dummy package that's never part of my dependency tree.)

this is on gitlab.
it creates e.g. 5 merge requests at the same instant.
as soon as i accept the first one, the other 4 are all in merge conflict.
How would it magically wake up and rebase these 4? You mean the next time i run it it would see the conflic, rebase, then i would accept 1 more, then it would rebase the next 3, and so on?
also, it would have the same merge conflict, how would it know how to resolve it?

**Comment by cjolowicz** [comment]
@donbowman I assumed you were referring to [GitHub Dependabot](https://docs.github.com/en/github/administering-a-repository/keeping-your-dependencies-updated-automatically). The bot recreates the PR (not a rebase, sorry for the sloppy terminology) when the changes conflict with the target branch.

**Comment by donbowman** [comment]
i see.
but it still leaves the original issue, that if i issue many update's, each as a single MR, after the first is accepted, the rest all conflict.
If the dependabot script is taught to then delete and recreate, ... its very slow.
if I were to run it daily, and had 10 updates, it would take 10 days to get them all in w/ this technique.

i guess, inefficiently, i could create some webhook that when the first mr is accepted, it could somehow run the dependabot again to recreate all the remaining updates, and then repeat every hour or so until all are done.

**Comment by seansfkelley** [comment]
It looks like Dependabot actually _does_ automatically rebase (well, [rewrite-force-pushes](https://github.com/dependabot/dependabot-core/issues/1722)) the PR by default, [according to the documentation](https://docs.github.com/en/github/administering-a-repository/configuration-options-for-dependency-updates#rebase-strategy). Judging by the docs and that issue, it sounds like it does it with webhooks or another timely mechanism rather than waiting for the next run, though neither one is explicit on how quick the turnaround is.

Maybe you're seeing merge conflicts triggered by `content-hash` because you've disabled this behavior?

**Comment by donbowman** [comment]
> It looks like Dependabot actually _does_ automatically rebase (well, [rewrite-force-pushes](https://github.com/dependabot/dependabot-core/issues/1722)) the PR by default, [according to the documentation](https://docs.github.com/en/github/administering-a-repository/configuration-options-for-dependency-updates#rebase-strategy). Judging by the docs and that issue, it sounds like it does it with webhooks or another timely mechanism rather than waiting for the next run, though neither one is explicit on how quick the turnaround is.
> 
> Maybe you're seeing merge conflicts triggered by `content-hash` because you've disabled this behavior?

content-hash conflicts even w/o dependabot. if 2 people change 2 branches, its in conflict.

the hosted dependabot of github may in fact have some faster trigger.
but the dependabot-core on a private gitlab does not. its a cron.

**Comment by lephuongbg** [comment]
My 2 cents, if the `content-hash` format can be structured as a list of sorted main dependency names and hashes calculated from resolved sub-dependencies, then it would reduce merge conflict chance a lot. Like

```toml
[metadata]
content-hash = [
    "astroid:03472c30eb2c53",
    "flask:bb564576db6a918",
    #...
]
```

**Comment by cjolowicz** [comment]
Here's an approach that has worked well for me and only uses git and (recent) Poetry:

```sh
git restore --staged --worktree poetry.lock
poetry lock --no-update
```

When rebasing a feature branch on main, this preserves pins from the main branch, and recomputes pins for your feature branch.

**Comment by AbdulRahmanAlHamali** [comment]
Most the answers here are answering what to do on a merge conflict, but we should be addressing why we're having a merge conflict in the first place. It is not reasonable to have a merge conflict whenever two developers modify different dependencies.  Having a conflict means you will need to realize that, merge master into your branch, recreate the lock file (which takes at least 3 minutes, because python...), re-push, re-run your CI (which could take a lot of time as well), etc. This is very bad UX.

@lephuongbg's comment is a possible solution. Is it possible to look into this?

Thank you a lot!

**Comment by adriangb** [comment]
This has become pretty problematic for us, to the point where it is _almost_ a dealbreaker. It would be great to find a way to restructure `poetry.lock` so that it can be more merge-friendly.

**Comment by nyanpasu64** [comment]
*(Note: I haven't touched Poetry in a while.)*

For prior art on designing merge-friendly lockfiles, Cargo's current iteration of lockfiles (eg. https://github.com/nyanpasu64/spectro2/blob/master/Cargo.lock) are flat and have neither a global content-hash nor a metadata array, but instead consisting of a large array of packages, storing metadata and checksums within each individual package's entry. The per-package checksum field might be for the same purpose as content-hash (though I'm not sure exactly what it does).

Also, cargo.lock files have changed format (https://github.com/rust-lang/cargo/pull/7070), going from a `[metadata]` table (as suggested in https://github.com/python-poetry/poetry/issues/496#issuecomment-734107507) to the current flat array, with the justification that it reduced the possibility for conflicts. Looking at [Cargo's release notes](https://github.com/rust-lang/cargo/blob/master/CHANGELOG.md), the new lockfile was introduced (disabled) in 1.38 (2019), made default in 1.41 (2020), but existing projects were not updated to "new format" lockfiles until 1.47 (2020).

**Comment by zyv** [comment]
I also think that the solution would be to make the lock file conflict free - even if resolving conflicts will be automated, updating will either require human interaction and/or disproportional amount of CI time.

**Comment by Mogost** [comment]
@sdispater could you provide your opinion on this issue, so that if someone takes the time to fix it they will have more information.

Since a PR (https://github.com/python-poetry/poetry/pull/2654) has been provided but it has not been reviewed, it is not clear if there is something that prevents having such an option. 

I also seem to have found a duplicate of this issue https://github.com/python-poetry/poetry/issues/4189

**Comment by finswimmer** [comment]
I believe most of suggestions here are thought to short. 

Getting somehow around the merge conflict due to the hash doesn't solve the problem. If there is a conflict, this means both sides changed the dependencies. Merging the locked dependencies doesn't necessarily results in a correct dependency tree.

The only clean solution I see right now, is the one suggested by @cjolowicz (https://github.com/python-poetry/poetry/issues/496#issuecomment-738680177)

**Comment by zyv** [comment]
> Getting somehow around the merge conflict due to the hash doesn't solve the problem.

@finswimmer it depends on which problem you are actually trying to solve.

You are right, conflict-free lockfile format wouldn't solve the problem of always producing a correct dependency tree (in some cases, arguably <<< 50% for fresh branches not deviating too much from master, the resulting tree will be incorrect).

It will however solve a very practical problem of having *most* merges produce a correct tree, which is what really matters in a typical branch-based development workflow. The broken trees will be immediately found by CI upon the next push, and so the suggestion by @cjolowicz will have to be applied (`poetry install` could detect this situation, output a text to this effect and suggest the commands to run to repair the lockfile).

This way the typical workflow will be "rebase / merge from master" and push, in 99% of the cases everything just works (tm) In some cases, `poetry install` in CI will detect a broken lockfile and ask to fix it with a few commands.

Currently the workflow is the opposite of that - and specifically to always do the @cjolowicz trick. It doesn't make the use of `poetry` impossible, just more annoying than it should be in my opinion.

**Comment by cjolowicz** [comment]
For those who encounter frequent merge conflicts in Dependabot PRs:

Dependabot updates version constraints in pyproject.toml even when the new version was already covered, see https://github.com/dependabot/dependabot-core/issues/4435. This means that Dependabot PRs for direct dependencies will *always* conflict with each other. One workaround for this limitation is the [lockfile-only versioning strategy](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates#versioning-strategy). If you have upper bounds on your version constraints, you will need to widen the constraints manually to receive major updates.

Note that there is a growing sentiment that upper version bounds are harmful in the Python ecosystem:

- [Should You Use Upper Bound Version Constraints?](https://iscinumpy.dev/post/bound-version-constraints/) and [Poetry Versions](https://iscinumpy.dev/post/poetry-versions/) by Henry Schreiner
- [Semantic Versioning Will Not Save You](https://hynek.me/articles/semver-will-not-save-you/) by Hynek Schlawack
- [Version numbers: how to use them?](https://bernat.tech/posts/version-numbers/) by Bernát Gábor
- [Why I don't like SemVer anymore](https://snarky.ca/why-i-dont-like-semver/) by Brett Cannon

Personally, replacing `^1.2.3`-style constraints with `>=1.2.3` and using the lockfile-only strategy has worked well for me. I rarely need to resolve merge conflicts, and Dependabot PRs are good for identifying breaking changes, including those that are not advertised in the version number.

**Comment by tony** [comment]
Issue should be renamed "auto-resolve _trivial_ / simple merge conflicts", or a new one should be created.  "all/most" is too much scope for a package manager.

When I have merge conflicts - which I've had hundreds of - it's often the content-hash, and at most, hashes and wheels.

This is extremely laborious and tedious.  

```
dependabot/pip/django-3.0.2 ❯ git pull --rebase origin master
From github.com:org/repo
 * branch                master     -> FETCH_HEAD
Auto-merging poetry.lock
CONFLICT (content): Merge conflict in poetry.lock
Auto-merging pyproject.toml
error: could not apply cadc1e314... :arrow_up: Bump django from 3.0.1 to 3.0.2
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply cadc1e314... :arrow_up: Bump django from 3.0.1 to 3.0.2
project on  HEAD (fae2bca) (REBASING 1/1) [=+] ❯ git diff
```

```diff
diff --cc poetry.lock
index cf2c1dcde,becb924fe..000000000
--- a/poetry.lock
+++ b/poetry.lock
@@@ -2897,7 -2896,7 +2897,11 @@@ testing = ["coverage (>=5.0.2)", "zope.
  [metadata]
  lock-version = "1.1"
  python-versions = "==3.*.*,>=3.8.1"
++<<<<<<< HEAD
 +content-hash = "889ccc59768f4c5a4c5dd14754e0b126d54f253171d583f0c63db16d601d6376"
++=======
+ content-hash = "9f357236d29da73cc3e57e6b48e2739bd154c1fc67cb130d7f6309a741575351"
++>>>>>>> cadc1e314 (:arrow_up: Bump django from 3.0.1 to 3.0.2)

  [metadata.files]
  aiohttp = [
```

**Comment by tony** [comment]
@cjolowicz I notice that https://github.com/cjolowicz/poetry-merge-lock is archived, any more details?  Maybe it'd be worth linking to potential workarounds or this PR in the README?

**Comment by cjolowicz** [comment]
@tony done

**Comment by tony** [comment]
@cjolowicz Thank you, this clarifies it for me! (as a workaround for the now)

**Comment by omBratteng** [comment]
As @tony mentioned, most merge conflicts that occur, are because of the content-hash in the lock file.
Because of these merge conflicts, if you have several PRs open from Dependabot to update several dependencies, you'll have to spend some time waiting between merges, since Dependabot has to rebase after each merge, just to update the content-hash.

**Comment by dimbleby** [comment]
The main use case that seems to have come up in this thread is dependabot, and there I am a little confused.  I don't see that anything has changed at the dependabot side since previous discussions: but I set up dependabot on a poetry project without any special instruction - and it already seems to update the lockfile only, when that's appropriate (eg https://github.com/dimbleby/e-treasure-hunt/pull/6).  

ie, at least in that example, dependabot already is making updates that would not cause merge conflicts.  

That aside, the basic ask of this issue seems to be that the `content_hash` be removed from `poetry.lock`.

An earlier [comment](https://github.com/python-poetry/poetry/issues/496#issuecomment-1002094619) said

> Getting somehow around the merge conflict due to the hash doesn't solve the problem. If there is a conflict, this means both sides changed the dependencies. Merging the locked dependencies doesn't necessarily results in a correct dependency tree.

This is all true, but the content hash is providing at best a partial defence: it is only doing anything for those cases where `pyproject.toml` changed.  If we had a series of commits like my dependabot one above we might experience no merge conflicts but still reach an invalid solution in the lockfile.

That is, I don't think that content-hash-as-defence-against-invalid-lockfiles is wholly convincing.

I also think it's useful to see that Rust's `Cargo.lock` does not include a similar hash - I wonder whether they thought about it and explicitly decided that such a hash was more trouble than it was worth, or just didn't.  If anyone has a link to discussion I'd be interested to see it.  "What would cargo do?" is often a pretty good way to think about such questions...

TLDR I'd like to hear what the case is for `content_hash` in the lockfile.

**Comment by colindean** [comment]
I'm using Makefiles in my project to capture tasks and I've now got these:

```Makefile
# FLAGS contains CPPFLAGS and LDFLAGS vars resolved from pkg-config
POETRY ?= $(FLAGS) poetry

.PHONY: fix-poetry-conflicts
fix-poetry-conflicts: ## Attempts to fix Poetry merge/rebase conflicts by choosing theirs and locking again
	git checkout --theirs poetry.lock
	$(MAKE) poetry-relock

.PHONY: fix-poetry-conflicts-2
fix-poetry-conflicts-2: ## Another way to try to fix Poetry merge/rebase conflicts
	git restore --staged --worktree poetry.lock
	$(MAKE) poetry-relock

.PHONY: poetry-relock
poetry-relock: pyproject.toml ## Run poetry lock w/o updating deps, use after changing pyproject.toml trivially
	$(POETRY) lock --no-update
```

Resolve the `pyproject.toml` conflicts first, then run one of the above depending on the situation.

**Comment by fredrikaverpil** [comment]
> TLDR I'd like to hear what the case is for content_hash in the lockfile.

@dimbleby did you come to any conclusions?

**Comment by finswimmer** [comment]
> TLDR I'd like to hear what the case is for content_hash in the lockfile.

The content_hash is the hash of relevant keys in the `pyproject.toml`. See https://github.com/python-poetry/poetry/blob/59a38bd935936e87aa6deae76764cedbdbca5d43/src/poetry/packages/locker.py#L254

Poetry needs this to detect if there were made changes to the `pyproject.toml` without changing  the lock file and give a warning to the user, that these file might be out of sync.

**Comment by samamorgan** [comment]
FYI, for anyone using Dependabot to keep versions updated, the best strategy I've found to speed things up is setting `open-pull-requests-limit: 1` on the pip config. This stops the cycle of: 5 PRs open -> One merges -> 4 PRs rebase

**Comment by djmattyg007** [comment]
@samamorgan I recommend using Renovatebot, which automatically groups updates.

**Comment by samamorgan** [comment]
> @samamorgan I recommend using Renovatebot, which automatically groups updates.

Thanks for the suggestion! This doesn't work for me in most projects. I want dependency updates to go through tests and if they fail, I want them to fail individually.

**Comment by jeffwidman** [comment]
FWIW from the Dependabot side it's now possible to configure `increase-if-necessary` as a version strategy for the python ecosystem, which can alleviate some of the pain here:
* https://github.com/dependabot/dependabot-core/issues/4435#issuecomment-975494765
* https://github.com/dependabot/dependabot-core/pull/5605

**Comment by domsj** [comment]
I've also noticed this problem on one of our projects.

It's not just dependabot needing to recreate PRs, it's also humans conflicting with each other and with dependabot.
And indeed everytime the human needs to do
```
git checkout --theirs poetry.lock
poetry lock --no-update
```
which is rather annoying.

> Poetry needs this to detect if there were made changes to the pyproject.toml without changing the lock file and give a warning to the user, that these file might be out of sync.

Perhaps instead of adding a hash, which always conflicts, poetry could add in the entire contents of pyproject.toml?
This way you get the desired check, and you get the exact same amount of conflicts as in pyproject.toml.
(Maybe there are even ways to improve on the idea by converting contents of pyproject.toml into some canonical form that has less chance of conflicting by adding in e.g. some extra lines - but would need to check git merge conflict detection more in detail for that)

Other thoughts:
- with this suggested adapted check one does not avoid having conflicting packages after merge - but that is what CI is for IMO
- if one does want the old behavior it could be opt-in (might want to configure it in pyproject.toml so that everyone automatically does the same, whether it's a human or dependabot)

**Comment by mrcljx** [comment]
Would it be possible to add a a config flag (e.g. tool config in `pyproject.toml`) that instructs Poetry to not render/expect the `content-hash` in the lock file for people who are running into merge conflicts a lot?

**Comment by neersighted** [comment]
I don't think so -- simply "don't add a hash" is actually quite dangerous. Like I said, the lock file is really a cache and not fully self-descriptive. Sure, when you remove the hash, you "solve" the merge conflicts, but that doesn't mean that the solution is correct (or up-to-date for your pyproject.toml).

I think we're better off devising a method to support mergable lock files as a first class feature, or explicitly disavowing removing the hash; adding a knob to opt-out is a really bad idea as it can already be done manually, and if we add a knob people will think it is supported, and become frustrated when Poetry produces incorrect results.

**Comment by Mogost** [comment]
So far, I'm getting frustrated with the amount of merger conflicts that are draining CI resources. This option may be appropriately named, that it is dangerous. This will solve the problem here and now and allow us to work on the hash problem as a whole.

**Comment by djmattyg007** [comment]
I recommend considering Renovatebot instead of Dependabot. It can combine multiple dependency updates into a single PR so that you don’t have to worry about having multiple conflicting PRs.

**Comment by Mogost** [comment]
However, I want to receive dependency updates separately. This clearly shows at what point the failure occurred.
And dependabot is not the only reason. People are in conflict with each other when there seems to be no conflict.

**Comment by mrcljx** [comment]
@neersighted I understand the suggestion is quite dangerous, but as more and more people are reporting that the conflict issue is turning into a major pain point, I was hoping that some kind of workaround could be made possible (+1 on giving this a very clear name).

---

Over the weekend I wanted to understand how other ecosystems (specifically Node.js) are solving the same problem (while keeping in mind that the [metadata cost] in Python is different). NPM stores the "self" package in the lock file as well (Yarn only does when using workspaces it seems):

```json5
// package-lock.json
{
 "name": "poetry-test",
 "version": "1.0.0",
 "lockfileVersion": 2,
 "requires": true,
 "packages": {
   "": {
     "name": "poetry-test",
     "version": "1.0.0",
     "license": "MIT",
     "dependencies": {
       "react": "^18.2.0"
     }
   },
```

Poetry instead stores a hash of `relevant_content` which more or less contains the dependencies of the package itself:

```
return sha256(json.dumps(relevant_content, sort_keys=True).encode()).hexdigest()
```

If we compare these two approaches, NPM's lockfile can be merged by git most of the time (and otherwise humans can merge it by hand) while Poetry is guaranteed to conflict when two commits modify the `pyproject.toml`.

While it seems quite possible to render `relevant_content` and not the `sha256` to the lock file (which actually sounds related to the ask of #1301), various comments in this thread have suggested this to be a feature to ensure/increase correctness:

- > Merging the locked dependencies doesn't necessarily results in a correct dependency tree.
- > but that doesn't mean that the solution is correct (or up-to-date for your pyproject.toml).

Generally, a git-merged lock file is not guaranteed to be "correct" regarding to the `pyproject.toml`. I'd love if someone could provide the thread with a non-esoteric case where updating `pyproject.toml` from two files that cleanly merges and not running `poetry lock` would cause a problem. I'm fairly certain that one exists, but I failed to contrive one.

That said, the correctness argument seems to be ignoring the fact that:

1. this problem exists in most other package managers as well.
   a. the solution to that is to check whether running the resolver again is a noop:
      1. see `npm ci`
      2. see `poetry lock --check`
4. AFAICT this seems to be true for other operations as well. The input to the algorithm is not just `pyproject.toml`, it is also the current lock file itself. `poetry update transitive-a` and `poetry update transitive-b` might also contradict each other, but the hash will not cause a conflict there.

The thing that makes (1) a bit unique in Poetry is that `poetry install` works quite different from other package managers. As an NPM user, you make a habit of running `node install` when pulling the main branch which will "heal" the lock file (and one definitely runs that after a merge conflict on the lock file). For Poetry, `poetry install` will not do that, instead new users must be made aware of the `lock` subcommand.

That said, this should more or less also work if Poetry stored the `relevant_contents` instead of the hash for the case that someone updated a version in `pyproject.toml` but then didn't run `poetry lock`.

Regarding (2) to me it seems that this hash is giving a false sense of correctness. The only true way to ensure things are correct is to run `poetry lock` habitually (or if possible run it on CI).

To summarize, I find it quite hard to defend the existence of the content-hash. (I personally believe that Poetry's UX would benefit if `install` performed a `lock --no-update`, but I understand that the metadata cost might be prohibitive here).

[metadata cost]: https://github.com/python-poetry/poetry/issues/496#issuecomment-1326725877

**Comment by neersighted** [comment]
> This option may be appropriately named, that it is dangerous. This will solve the problem here and now and allow us to work on the hash problem as a whole.

This is easy to assert from a non-maintainer standpoint, but we already pay a very high maintenance/support burden for other poorly designed/second-class features. I suspect you will not be able to find any consensus to merge another one.

> Regarding (2) to me it seems that this hash is giving a false sense of correctness. The only true way to ensure things are correct is to run poetry lock habitually (or if possible run it on CI).

You're not incorrect, but this is not the intended way to use Poetry/a primary workflow. In general what is intended is `poetry update,` run CI, and determine if any failures result from broken versions, at which point they are excluded from the constraints in pyproject.toml.

If we drop hashing and make the lock file self-descriptive, we'll greatly increase the maintenance burden for the lock file, but that may be a price worth paying. However, we make the common case much more risky/dangerous. People are going to blindly trust Dependabot/Git to update the lock file, and will be opening bugs against Poetry when they get unexpected results. I'm not sure that "make sure you run `poetry lock --no-update` before you open a bug will fly..."

I'd still like to hear what some of the SMEs in this area think, but since we're now talking about a "no-hash" approach instead of trying to change the solver, I'll also ping @abn @finswimmer @sdispater for opinions.

**Comment by neersighted** [comment]
A final thought: we can never come up with the 'canonical' (what the solver would pick, given all information was available to it) solution to lock file merges, but maybe we can at least start validating the lock file makes sense? That is to say, we store the dependency tree, so we could at least validate that it still makes sense/the solution doesn't have internal conflicts. We might already do that, but it certainly doesn't happen in the parts of the installer I am familiar with... cc @radoering @dimbleby

**Comment by radoering** [comment]
> 2\. see `poetry lock --check`

This command just checks for the correct hash. (That's why it is quite fast.)

> (I personally believe that Poetry's UX would benefit if `install` performed a `lock --no-update`, but I understand that the metadata cost might be prohibitive here).

I disagree. Even for large projects `poetry install` is quite fast if most of the packages are already installed. `poetry lock --no-update` takes significantly longer and it's just not necessary if you have taken precautions to make sure that the lock file checked into your repo is alright.

> We might already do that, but it certainly doesn't happen in the parts of the installer I am familiar with...

More or less. Actually, each install contains a solver run for the target environment with the lockfile being the only available repository. If the lockfile is not consistent with the pyproject.toml you get some sort of a solver error many users don't understand. The only hint for the reason of these errors is the warning about an inconsistent lock file (based on the hash). Just a quick idea, but we could derive "solver error during install" -> "inconsistent lock file". (It's probably not that easy because there are probably a bunch of different exceptions we have to consider but it might work.) Anyway, it's less expensive to validate the lockfile for the current environment (during install) than to validate it in general.

**Comment by neersighted** [comment]
> The only hint for the reason of these errors is the warning about an inconsistent lock file (based on the hash). Just a quick idea, but we could derive "solver error during install" -> "inconsistent lock file". (It's probably not that easy because there are probably a bunch of different exceptions we have to consider but it might work.) Anyway, it's less expensive to validate the lockfile for the current environment (during install) than to validate it in general.

Right, this matches my current recollection. If it's a full solver run (and not abbreviated like I thought), then it's more robust than I realized. The missing piece is inferring that a solver error during install is likely caused by an inconsistent lock file; I wonder if re-trying with an automatic re-lock makes sense in that case?

**Comment by dimbleby** [comment]
`poetry lock --no-update` is fast when the relevant part of the poetry cache is populated - but that's not usually the case in CI pipelines.  Then that command has to go and retrieve metadata from pypi or other repositories (which can indeed be slow).  For similar reasons `poetry install` is probably slow in CI pipelines.

---

In general it is not possible safely to merge two independent changes to a lockfile without re-solving.  One can construct examples in which it an upgrade to A is possible, an upgrade to B is possible, but the upgraded A and B are incompatible with one another.  So if some bot has proposed both upgrades then it would be an error to merge both, even if they are compatible so far as git can tell.  In that sense, being merge-friendly would actually be an anti-feature!

---

That being said, I still am confused as to what problem folk are actually seeing here.  eg my experience with dependabot is that it updates only the lockfile, not pyproject.toml, and therefore its updates never make changes to the hash in the lockfile.  Clearly there's something going on that y'all want to talk about - but it's not what I'm seeing!

**Comment by radoering** [comment]
> I wonder if re-trying with an automatic re-lock makes sense in that case?

I wouldn't want have an automatic re-lock. IMO, `install` should not change the lock file. At least not per default.

**Comment by neersighted** [comment]
> In general it is not possible safely to merge two independent changes to a lockfile without re-solving.

This has always been my general objection. However, I wonder if users might find it more desirable to use a "best-guess" (Git merge-based) solution when possible, and to bail out when an incompatible merge happened (since we in theory have the metadata to detect that/already run the solver). 

However, I just noticed a seeming non-sequitur from @Mogost:

> So far, I'm getting frustrated with the amount of merger conflicts that are draining CI resources.

I don't understand what you mean -- a conflicted `poetry.lock` is actually going to use less resources to detect than a "merged" lock file that is not a usable solution. So in that sense, I don't think you'll get what you want if we "solve" this (which appears to be possible by choosing to compromise/break at a different point, at least to me).

**Comment by dimbleby** [comment]
I've been looking more carefully at my dependabot MRs to understand why sometimes dependabot updates pyproject.toml and sometimes it is content simply to update poetry.lock

I _think_ the difference is that `>=` constraints don't cause it to update pyproject.toml, but `^` constraints do.

So perhaps there's some code already in dependabot that is trying to be smart about this, but which could be taught about `^`.  That would likely significantly reduce the amount of noise in the hash value.

Maybe someone on this thread who finds this bothersome would like to raise an issue (or merge request) at dependabot to investigate (or fix) this.

**Comment by djmattyg007** [comment]
Even if you don’t want renovate to group dependency updates, you can still do things like ask it to only create one PR at a time, for example. This can reduce the burden of wasted CI time. No matter which way you look at it, it will still make your life easier in dealing with this poetry lock file problem.

**Comment by zyv** [comment]
> However, I just noticed a seeming non-sequitur from @Mogost:
> 
> > So far, I'm getting frustrated with the amount of merger conflicts that are draining CI resources.
> 
> I don't understand what you mean -- a conflicted `poetry.lock` is actually going to use less resources to detect than a "merged" lock file that is not a usable solution.

Well, it's a **non**-non-sequitur.

Right now, if you want a completely automated solution to resolve merge conflicts with Poetry lockfiles, you have to tell dependency bots to "rebase" their PRs every time a PR touching the lockfile is merged. I'm putting "rebase" in quotes, because in the context of dependency bots like dependabot or Renovate it's not about git rebase, but rather about re-locking the lockfile.

So, imagine, you have 4 dependabot PRs. You merge 1 PR, and trigger 3 PRs to re-lock and re-run the CI (not only Poetry, which might take an insignificant amount of time as compared to tests, but everything). Then you merge the 2nd one and the dance repeats.

This is what drains CI-time, not a possibly failing Poetry call, which (if it indeed fails) will actually conserve CI time, because downstream jobs just won't be able to proceed.

So I think your thinking goes into the right direction in terms of saving CI time.

**Comment by neersighted** [comment]
Maybe this is a matter of differing workflows, but as when I use automated tools there is one PR or they are batched, the prospect of "random" failures requiring human intervention during the install stage seems much more likely to waste CI time.

**Comment by mrcljx** [comment]
The content-hash remains to be a big point of frustration. When three engineers add dependencies it's a constant race who gets their merge in first. The other two will have to rebase either once or twice to fix the `content-hash`.

**Comment by simonpercivall** [comment]
The issue I'm most afraid over right now over how the `poetry.lock` file works is updating sub-dependencies that don't touch `pyproject.toml`. Now, in a pretty big team of mixed-seniority devs, they don't always know how to properly merge and inspect the result, and it's happened more than once that an upgraded sub-dependency gets _invisibly_ downgraded in a merge conflict resolution (even worse, git makes it _really_ hard to see things that change in a merge commit); that is, the dev blindly checks out `--theirs` or `--ours`, and then runs `poetry lock --no-update`.

If there was even an imperfect way poetry could _decrease_ the surface of a conflict (and just having a built-in way to resolve `content-hash` conflicts would solve like 95% of conflicts), it would go a long way to making it _easier_ to use `poetry` in a team. (Sure, we could go back to pinning every sub-dependency in `pyproject.toml`, but then what's the point.)

In a tool that I otherwise _really_ enjoy using, this is a painful pain point.

**Comment by dimbleby** [comment]
> The issue I'm most afraid over right now over how the poetry.lock file works is updating sub-dependencies that don't touch pyproject.toml

since there is no content-hash change when pyproject.toml is unchanged,  a "built-in way to resolve `content-hash` conflicts" would make exactly no difference in these cases

**Comment by simonpercivall** [comment]
> > The issue I'm most afraid over right now over how the poetry.lock file works is updating sub-dependencies that don't touch pyproject.toml
> 
> since there is no content-hash change when pyproject.toml is unchanged, a "built-in way to resolve `content-hash` conflicts" would make exactly no difference in these cases

Ah, I explained badly:
1. In Branch A Update a sub dependency
2. In A Update the `pyproject.toml`
3. In branch B, update `pyproject.toml`
4. In B, merge A, resolve the conflict _badly_, `lock --no-update`, steep (1) is gone

But as been said earlier in the discussion: Of course, to go as far as possible, any solution would need to look at both sides of the merge, and try to solve them together (like the NPM way someone described, or like the now-outdated "poetry-merge-file" package). And this is of course much harder to maintain.

**Comment by simonpercivall** [comment]
> This has always been my general objection. However, I wonder if users might find it more desirable to use a "best-guess" (Git merge-based) solution when possible, and to bail out when an incompatible merge happened (since we in theory have the metadata to detect that/already run the solver).

@neersighted Have you thought any more about this idea? To me, it sounds ideal if possible.

**Comment by king-of-poppk** [comment]
This is not just a dependabot issue: if you work on a long-lived branch where you ad/remove/upgrade dependencies good luck with rebasing against external dependency upgrades (for instance, merged dependabot PRs)!

**Comment by Andrew-Chen-Wang** [comment]
This is more for myself, but having to rebase all the time, just accept the incoming rebase:

```
git rebase -s recursive -X origin/main
```

then re-lock all the lock files. This primarily assumes the development branch's PR is JUST for adding a new package.

**Comment by SlowMo24** [comment]
As far as I see this is mainly about the hash and the content of the `poetry.lock`. I'm annoyed by a smaller issue: Putting the poetry version into the lock-file header basically creates an edit war in our repository:

`# This file is automatically @generated by Poetry 1.8.2 and should not be changed by hand.`

By recording the full version, each developer will change that line if they are on a slightly different version. I suggest to either only record versions that are relevant for the file (i.e. incompatible in creating or reading it) or not recording the version at all.

**Comment by webknjaz** [comment]
PEP 751 was accepted yesterday. Poetry should implement that instead.

**Comment by king-of-poppk** [comment]
> PEP 751 was accepted yesterday. Poetry should implement that instead.

See: https://peps.python.org/pep-0751/#hashing-the-contents-of-the-lock-file-itself.
